### ISSUE TITLE
Rendering wrong post in responses?

### ISSUE THREAD

**Issue by kameko** [lobsters/lobsters#1313]
![Screenshot 2024-08-27 221936](https://github.com/user-attachments/assets/46a63866-1655-4d65-bd43-85af04d99b49)

I have seen this happen a lot today. The two posts are *not* responding to the post they are under, but to the comment above the one they appear to be responding to, they have been switched. It seems to resolve itself after a little while, but I've seen this several times today. Something seems seriously wrong with how posts are sorted/rendered, even after I perform a browser cache refresh.

**Comment by pushcx** [comment]
#1308 touched this code and was deployed yesterday, so it’s almost certainly the culprit.

**Comment by pushcx** [comment]
If it's #1308 the missing test that would've caught it is basically the existing test that `update_and_recalculate_score!` doesn't change the `confidence_order` but on a reply, not a top-level comment.

**Comment by roryokane** [comment]
I am seeing this bug too on https://lobste.rs/s/7pvs12/hurl_5_0_0_parallel_edition. The parent comment of [square_usual‘s comment](https://lobste.rs/s/7pvs12/hurl_5_0_0_parallel_edition#c_1hawhn) sometimes switches between two particular comments after I refresh the page:

### Screenshot at 11:48:40 AM

square\_usual’s comment has an incorrect parent. It is obviously incorrect because the timestamps and unread status don’t match.

> <img width="1271" alt="Lobsters story 7pvs12 screenshot 2024-08-31 at 11 48 40 AM" src="https://github.com/user-attachments/assets/700ca3f5-0466-4297-a629-1815873b55ab">

### Screenshot at 11:53:19 AM

square\_usual’s comment has the original, correct parent.

> <img width="1163" alt="Lobsters story 7pvs12 screenshot 2024-08-31 at 11 53 19 AM" src="https://github.com/user-attachments/assets/8e5ae114-d8ed-49e5-9d94-48fef7e22427">

### Screenshot at 11:54:14 AM

After refreshing about five times, I saw square\_usual’s comment have the incorrect parent again.

> <img width="1193" alt="Lobsters story 7pvs12 screenshot 2024-08-31 at 11 54 14 AM" src="https://github.com/user-attachments/assets/66aad3cb-cae2-4c66-ab21-8acc4fa565db">

**Comment by pushcx** [comment]
I realized I should've pinged @byroot when I mentioned his PR. I don't
have a stable internet connection right now but I plan to revert when I
can.

**Comment by byroot** [comment]
Arf. Happy to look some more into this in the future if you got a failing test. Was a nice optimization :/

**Comment by pushcx** [comment]
No, but I have had two more reports like the ones in this thread. I am
not really able to write code for a few days. I suspect it is something
about the comments being replies instead of top-level.

**Comment by sknebel** [comment]
another report: https://lobste.rs/s/zyketc/current_zlib_rs_performance#c_otsifm

**Comment by BenjaminRi** [comment]
Yes, this happened to me just now. Something really strange is going on:

Sometimes it looks like this:

![Screenshot_20240903_195713](https://github.com/user-attachments/assets/5bd4228d-f015-4fda-aafa-51ef5821c39d)

When I refresh, it looks like this:

![Screenshot_20240903_195822](https://github.com/user-attachments/assets/fa6d053e-7e02-4d65-9b22-b076c98c58b3)

I clearly remember that I replied to the correct comment (I did a minor edit in the second screenshot). The Hans Boehm comment also gets mixed up all the time.

**Comment by pushcx** [comment]
This was a bug in #1308, which I reverted in 4ed2d708. Having duplicate `confidence_order` values meant that replies were getting parented under the wrong commit.

After attempting a couple alternate approaches, we realized that there were further problems, where Story caches were not getting updated until any comment on the story got a vote, busting that cache. This is especially bad for stories going from 0 comments to 1 comment because the `comments_cache` column wasn't updating, so visitors wouldn't see there's a comment to go vote on it.

@byroot I appreciate your attempt in #1308 at minimizing the number of queries for comment creation, but these bugs meant it was worth reverting, at least for now. Maybe a different approach would be better; I've considered removing our reliance on ActiveRecord lifecycle callbacks to use something like a service method so all this logic is apparent rather than spread throughout the models.

There's a longer discussion of this in [today's stream](https://push.cx/stream/2024-09-05). (link live in a couple minutes)

**Comment by kameko** [comment]
Is this change deployed yet? Because these responses do not look like they are responding to the relevant post.

![Screenshot 2024-09-05 193112](https://github.com/user-attachments/assets/6e28f8c2-3448-486c-abcc-07946481539c)

**Comment by pushcx** [comment]
I _thought_ it was, but some of the other code I touched after didn't go live. I've just run another deploy so it should certainly be live now.

I didn't recalc all the existing scores, so I did that now (`Comment.where(created_at: 11.days.ago..).find_each {|c| c.update_score_and_recalculate! 0,0 }`). But that isn't the issue in your screenshot, given the times of the comments. I had been waiting on #1318 that I've been discussing with @dzwdz all day; they've found at least one remaining bug in the comment sorting and that/those will prompt another recalc.

I've looked at the comments you screenshotted about an hour later and they've all got more votes in the same position, so this caching bug couldn't be at work. I think maybe these are just particularly disjointed comments. proctrap is quoting [the comment](https://lobste.rs/s/yx57uf/is_linux_collapsing_under_its_own_weight#c_dptfuw) that lpil is responding to (just above the screenshot) build on them both; wrs is elaborating on what makes code hard to read. deepchasm is trying to say they left a comment and then thought better of it and removed it; I checked the http logs to confirm. lonjil is quoting and asking about the original comment that deepchasm removed.

I think we're OK for now, but please don't hesitate to reopen the issue or reach out to me if you see something else that looks like this bug recurring. The really frustrating thing about "sometimes" bugs like this is that it takes a while to rebuild trust that they're truly fixed!

**Comment by shrik450** [comment]
Saw this again today:

<img width="821" alt="image" src="https://github.com/user-attachments/assets/38f16097-8e01-4b1f-b1d4-17d82bdc21f9">

Link to the misplaced comment: https://lobste.rs/s/efkk2t/what_are_you_doing_this_week#c_om0ixf
